File: src/main/java/io/github/boogiemonster1o1/cartses/Cartses.java
package io.github.boogiemonster1o1.cartses;

import net.fabricmc.api.ModInitializer;

import io.github.boogiemonster1o1.cartses.entity.ModEntityTypes;
import io.github.boogiemonster1o1.cartses.item.ModItems;
import org.apache.logging.log4j.Level;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

public class Cartses implements ModInitializer {
	public static Logger LOGGER = LogManager.getLogger();

	public static final String MOD_ID = "cartses";

	@Override
	public void onInitialize() {
		log(Level.INFO, "Initializing");
		ModEntityTypes.init();
		ModItems.init();
	}

	public static void log(Level level, String message) {
		LOGGER.log(level, "[" + Cartses.class.getSimpleName() + "] " + message);
	}
}

File: src/main/scala/io/github/boogiemonster1o1/cartses/Cartses.scala
/*
 * Copyright (c) 2020 BoogieMonster1O1
 *
 * Permission to use, copy, modify, and/or distribute this software for
 * any purpose with or without fee is hereby granted, provided that the
 * above copyright notice and this permission notice appear in all
 * copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL
 * WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE
 * AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL
 * DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR
 * PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER
 * TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
 * PERFORMANCE OF THIS SOFTWARE.
 */

package io.github.boogiemonster1o1.cartses

import io.github.boogiemonster1o1.cartses.entity.ModEntityTypes
import io.github.boogiemonster1o1.cartses.entity.cart.MinecartWithNoteBlockEntity
import io.github.boogiemonster1o1.cartses.entity.networking.EntityUtils
import io.github.boogiemonster1o1.cartses.item.ModItems
import net.fabricmc.api.ModInitializer
import net.fabricmc.fabric.api.network.{PacketContext, ServerSidePacketRegistry}
import net.minecraft.network.PacketByteBuf
import org.apache.logging.log4j.{Level, LogManager, Logger}

object Cartses extends ModInitializer {
	override def onInitialize(): Unit = {
		log(Level.INFO, "Initializing Cartses")
		ModEntityTypes.init()
		ModItems.init()
		ServerSidePacketRegistry.INSTANCE.register(EntityUtils.updateScreenC2SId, (ctx: PacketContext, buf: PacketByteBuf) => {
			val entitySyncId: Int = buf.readVarInt()
			val note: Int = buf.readVarInt()
			val repeat: Boolean = buf.readBoolean()
			ctx.getTaskQueue.execute(() => {
				val entity: MinecartWithNoteBlockEntity = ctx.getPlayer.world.getEntityById(entitySyncId).asInstanceOf[MinecartWithNoteBlockEntity]
				entity.setNote(note)
				entity.setRepeat(repeat)
			})
		})
	}

	val modId = "cartses"
	val logger: Logger = LogManager.getLogger

	def log(level: Level, message: String): Unit = {
		logger.log(level, "[Cartses] " + message)
	}
}
